R/scalar_in_linter.R

```diff
@@ -0,0 +1,42 @@
+#' Block usage like x %in% "a"
+#'
+#' `vector %in% set` is appropriate for matching a vector to a set, but if
+#'   that set has size 1, `==` is more appropriate. `%chin%` from `data.table`
+#'   is matched as well.
+#'
+#' `scalar %in% vector` is OK, because the alternative (`any(vector == scalar)`)
+#'   is more circuitous & potentially less clear.
+#'
+#' @evalRd rd_tags("scalar_in_linter")
+#' @seealso [linters] for a complete list of linters available in lintr.
+#' @export
+scalar_in_linter <- function() {
+  # TODO(#2085): Extend to include other cases where the RHS is clearly a scalar
+  # NB: all of logical, integer, double, hex, complex are parsed as NUM_CONST
+  xpath <- "
+  //SPECIAL[text() = '%in%' or text() = '%chin%']
+    /following-sibling::expr[NUM_CONST or STR_CONST]
```

Comment:
We should either exclude NAs from this XPath or implement the branching right away I think. What do you think about it?

Reply:
Oh, I missed your point earlier I think -- the recommendation to use `==` is just wrong for NA on the RHS (it's an `equals_na_linter()` violation too). I think the reason this escaped us so far is we'd already fixed the handful of `equals_na_linter()` hits in our codebase by the time this was written.

Agree it should be addressed up-front.

Reply:
Done. For now, we customize the message for `NA` on the RHS. More parameterization to follow in subsequent enhancements.